php/src/CustomAuthentication.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace Ionic;
+
+use Firebase\JWT\JWT;
+
+class CustomAuthentication {
+    /**
+     * @param string GET parameter token.
+     * @param string GET parameter state.
+     * @return string Redirect URI.
+     * @throws \Exception
+     */
+    public static function process($token, $state) {
+        $request = JWT::decode($token, static::SECRET, ["HS256"]);
+
+        static::validateResponse($request);
+
+        $credentials = [
+            'username' => 'dan',
+            'password' => '123',
+            'user_id' => 1
+        ];
+
+        // TODO: Authenticate your own real users here
+        if ($request->data->username != $credentials['username'] || $request->data->password != $credentials['password']) {
+            throw new \Exception('Invalid credentials.');
+        } else {
+            $user_id = $credentials['user_id'];
+        }
+
+        $token = JWT::encode(['user_id' => $user_id], static::SECRET);
+
+        $redirect_uri = $_GET['redirect_uri'].'&token='.urlencode($token).'&state='.$state;
```

Comment:
True, it doesn't seem necessary, there are more parameters in the redirect_uri so you're not building the whole query just appending two parameters which this approach shows clearly. There's a lot of optimization that could be added but I think this is a clear example.